### PR TITLE
GVL release

### PR DESCRIPTION
**Pull request by atesgoral**
Let Faiss index search release GVL while doing the CPU-intensive search, so maybe multi threads can benefit.

```
dev clone faiss-ruby
git submodule update --init
dev up
bundle exec rake compile
bundle exec rake test TEST=test/thread_benchmark.rb
```

### CODE REVIEW

**Comment by atesgoral** [comment]
Without changes, with 50 threads:

```rb
    nb = 100000 # number of vectors in the database
    nq = 10 # number of queries
```

```
Time taken: 3.9595079999417067 seconds
Time taken with threads: 4.061103000072762 seconds
```

**Comment by atesgoral** [comment]
After incorporating `rb_thread_call_without_gvl`:

```
Time taken: 3.906453000032343 seconds
Time taken with threads: 4.774259999976493 seconds
```

`¯\_(ツ)_/¯`

File: ext/faiss/index.cpp [hunk 1]
```diff
@@ -136,7 +165,17 @@ void init_index(Rice::Module& m) {
...
+        index_search_args args = {
+          .index = &self,
+          .n = n,
+          .x = objects.read_ptr(),
+          .k = k,
+          .distances = distances.write_ptr(),
+          .labels = labels.write_ptr(),
+        };
+
+        rb_thread_call_without_gvl(index_search_func, &args, RUBY_UBF_IO, NULL);
```
Comment (atesgoral) [inline ext/faiss/index.cpp]: Should this be `RUBY_UBF_IO` or `NULL`?

File: Rakefile [hunk 2]
```diff
@@ -11,6 +11,7 @@ end
 Rake::ExtensionTask.new("faiss") do |ext|
   ext.name = "ext"
   ext.lib_dir = "lib/faiss"
+  ext.config_options = "--with-cppflags=-I#{RbConfig::CONFIG['rubyhdr']}"
```
Comment (atesgoral) [inline Rakefile]: IDK if there's cleaner / more Rake::ExtensionTaskic way of doing this.

**Comment by atesgoral** [comment]
Is Rice getting in the way because it has some thread safety features built into it?

File: ext/faiss/index.cpp [hunk 3]
```diff
@@ -90,6 +92,33 @@ namespace Rice::detail {
...
+ * return at most k vectors. If there are not enough results for a
+ * query, the result array is padded with -1s.
+ *
+ * @param n           number of vectors
+ * @param x           input vectors to search, size n * d
+ * @param k           number of extracted vectors
+ * @param distances   output pairwise distances, size n*k
+ * @param labels      output labels of the NNs, size n*k
+ */
+struct index_search_args {
```
Comment (atesgoral) [inline ext/faiss/index.cpp]: Maybe I can use a lambda instead of having a wrapper function

File: test/thread_benchmark.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      50.times do
+        index.search(xq, k) # search for the nearest neighbors
+      end
+    }
+
+    puts "Time taken: #{time} seconds"
+
+    time = measure {
+      (0...50).map do
+        Thread.new { index.search(xq, k) }
```
Comment (atesgoral) [inline test/thread_benchmark.rb]: Will try pre-creating / pooling

File: ext/faiss/index.cpp (hunk 1 above)
Comment (tenderlove) [reply ext/faiss/index.cpp]: Is this code doing IO or CPU?

Comment (atesgoral) [reply ext/faiss/index.cpp]: It's gotta be doing pure CPU. We populate an in-memory store with a shit ton of vectors and then all "search" should be doing is doing some vector distance + additional math over huge sets of numbers.

Comment (atesgoral) [reply ext/faiss/index.cpp]: I used `RUBY_UBF_IO` because Cursor told me to. I bow to our AI overlords.

Comment (tenderlove) [reply ext/faiss/index.cpp]: I think we want [`RUBY_UBF_PROCESS`](https://github.com/ruby/ruby/blob/7be9a333cabd97a17a2926b15f756f2ef9e57243/thread.c#L1594-L1595)

Comment (atesgoral) [reply ext/faiss/index.cpp]: :<

```
Time taken: 4.434523000032641 seconds
Time taken to create threads: 0.00047199998516589403 seconds
Time taken with threads: 5.277729999972507 seconds
```

**Comment by atesgoral** [comment]
🎉 
```
Running single-threaded search...
Time taken: 5.406318999943323 seconds
Running multi-threaded search with 2 threads...
Time taken with threads: 2.582810999942012 seconds
```

**Comment by atesgoral** [comment]
Closing in favour of https://github.com/ankane/faiss-ruby/pull/7

**Pull request closed without merging**